File: scripts/providers/get_agy_project_sessions.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Antigravity (AGY) Project Session Provider
Scans ~/.gemini/antigravity/brain for sessions matching the specified project root,
extracting recent activities, touched files, and candidate functional tags.
"""

import sys
import os
import json
import re
from pathlib import Path
from datetime import datetime, timezone


def normalize_path(path_str: str) -> str:
    if not path_str:
        return ""
    return str(Path(path_str).resolve()).replace("\\", "/").rstrip("/").lower()


def scan_agy_sessions(project_root_str: str = ".", custom_brain_path: str = None, inspect_activity: bool = True) -> list:
    project_root = normalize_path(project_root_str)
    
    if custom_brain_path:
        brain_dir = Path(custom_brain_path)
    else:
        brain_dir = Path.home() / ".gemini" / "antigravity" / "brain"
        
    if not brain_dir.is_dir():
        return []

    sessions = []
    
    for entry in brain_dir.iterdir():
        if not entry.is_dir():
            continue
        conv_id = entry.name
        
        # Check transcript log
        log_file = entry / ".system_generated" / "logs" / "transcript.jsonl"
        if not log_file.is_file():
            log_file = entry / ".system_generated" / "logs" / "transcript_full.jsonl"
            if not log_file.is_file():
                continue

        try:
            stat = log_file.stat()
            last_modified = datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc).isoformat()
            created = datetime.fromtimestamp(stat.st_ctime, tz=timezone.utc).isoformat()
        except Exception:
            last_modified = ""
            created = ""

        is_match = False
        title = f"Session {conv_id}"
        user_prompts = []
        touched_files = set()

        try:
            with open(log_file, "r", encoding="utf-8", errors="ignore") as f:
                for line in f:
                    line_lower = line.lower()
                    if not is_match and (project_root in line_lower or project_root.replace(":", "%3a") in line_lower):
                        is_match = True
                    
                    # Extract user input prompts
                    if '"type":"USER_INPUT"' in line or '"type": "USER_INPUT"' in line:
                        match = re.search(r'"content"\s*:\s*"([^"]+)"', line)
                        if match:
                            raw_content = match.group(1)
                            clean_content = re.sub(r'<[^>]+>', '', raw_content)
                            clean_content = clean_content.replace("\\n", " ").replace('\\"', '"').strip()
                            if clean_content and clean_content not in user_prompts:
                                user_prompts.append(clean_content[:150])
                    
                    # Extract touched files from tool calls if inspect_activity enabled
                    if inspect_activity:
                        file_matches = re.findall(r'"(?:AbsolutePath|TargetFile|SearchPath)"\s*:\s*"([^"]+)"', line)
                        for fm in file_matches:
                            clean_fm = fm.replace('\\\\', '/').replace('\\', '/')
                            if clean_fm.lower().startswith(project_root):
                                rel_path = clean_fm[len(project_root):].lstrip('/')
                                if rel_path:
                                    touched_files.add(rel_path)
                            elif not clean_fm.startswith(('http://', 'https://')):
                                touched_files.add(clean_fm.split('/')[-1])
        except Exception:
            pass

        if is_match:
            if user_prompts:
                title = user_prompts[0]
                if len(title) > 60:
                    title = title[:57] + "..."

            agents_md = Path(project_root_str) / "AGENTS.md"
            rule_files = [str(agents_md.resolve()).replace("\\", "/")] if agents_md.is_file() else []
            
            session_data = {
                "vendor": "antigravity",
                "session_id": conv_id,
                "title": title,
                "project_root": project_root,
                "is_active": True,
                "created_at": created,
                "last_active_at": last_modified,
                "log_path": str(log_file.resolve()).replace("\\", "/"),
                "rule_files": rule_files
            }

            if inspect_activity:
                session_data["recent_prompts"] = user_prompts[-5:] if len(user_prompts) > 5 else user_prompts
                session_data["recent_touched_files"] = sorted(list(touched_files))[:15]

            sessions.append(session_data)

    # Sort by last_active_at descending
    sessions.sort(key=lambda s: s["last_active_at"], reverse=True)
    return sessions


def main():
    if hasattr(sys.stdout, "reconfigure"):
        try:
            sys.stdout.reconfigure(encoding="utf-8")
        except Exception:
            pass
    project_root = sys.argv[1] if len(sys.argv) > 1 else "."
    inspect = "--inspect" in sys.argv or "-i" in sys.argv
    sessions = scan_agy_sessions(project_root, inspect_activity=inspect)
    print(json.dumps(sessions, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/providers/get_agy_project_sessions.js
#!/usr/bin/env node
/**
 * Antigravity (AGY) Project Session Provider (Node.js)
 * Scans ~/.gemini/antigravity/brain for sessions matching the specified project root.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

function normalizePath(p) {
  if (!p) return '';
  return path.resolve(p).replace(/\\/g, '/').replace(/\/+$/, '').toLowerCase();
}

function scanAgySessions(projectRootStr = '.', customBrainPath = null, inspectActivity = true) {
  const projectRoot = normalizePath(projectRootStr);
  const brainDir = customBrainPath ? path.resolve(customBrainPath) : path.join(os.homedir(), '.gemini', 'antigravity', 'brain');

  if (!fs.existsSync(brainDir) || !fs.statSync(brainDir).isDirectory()) {
    return [];
  }

  const sessions = [];
  const entries = fs.readdirSync(brainDir);

  for (const entry of entries) {
    const entryPath = path.join(brainDir, entry);
    if (!fs.statSync(entryPath).isDirectory()) continue;

    const convId = entry;
    let logFile = path.join(entryPath, '.system_generated', 'logs', 'transcript.jsonl');
    if (!fs.existsSync(logFile)) {
      logFile = path.join(entryPath, '.system_generated', 'logs', 'transcript_full.jsonl');
      if (!fs.existsSync(logFile)) continue;
    }

    let created = '';
    let lastModified = '';
    try {
      const stat = fs.statSync(logFile);
      created = stat.birthtime ? stat.birthtime.toISOString() : stat.ctime.toISOString();
      lastModified = stat.mtime.toISOString();
    } catch (e) {
      // ignore
    }

    let isMatch = false;
    let title = `Session ${convId}`;
    const userPrompts = [];
    const touchedFiles = new Set();

    try {
      const content = fs.readFileSync(logFile, 'utf8');
      const lines = content.split('\n');

      for (const line of lines) {
        if (!line.trim()) continue;
        const lineLower = line.toLowerCase();
        const projectRootEscaped = projectRoot.replace(':', '%3a');

        if (!isMatch && (lineLower.includes(projectRoot) || lineLower.includes(projectRootEscaped))) {
          isMatch = true;
        }

        // Extract user prompts
        if (line.includes('"type":"USER_INPUT"') || line.includes('"type": "USER_INPUT"')) {
          const match = line.match(/"content"\s*:\s*"([^"]+)"/);
          if (match) {
            let rawContent = match[1];
            let clean = rawContent.replace(/<[^>]+>/g, '').replace(/\\n/g, ' ').replace(/\\"/g, '"').trim();
            if (clean && !userPrompts.includes(clean)) {
              userPrompts.push(clean.substring(0, 150));
            }
          }
        }

        // Extract touched files
        if (inspectActivity) {
          const fileMatches = line.matchAll(/"(?:AbsolutePath|TargetFile|SearchPath)"\s*:\s*"([^"]+)"/g);
          for (const fm of fileMatches) {
            const cleanFm = fm[1].replace(/\\\\/g, '/').replace(/\\/g, '/');
            if (cleanFm.toLowerCase().startsWith(projectRoot)) {
              const rel = cleanFm.substring(projectRoot.length).replace(/^\/+/, '');
              if (rel) touchedFiles.add(rel);
            } else if (!cleanFm.startsWith('http://') && !cleanFm.startsWith('https://')) {
              touchedFiles.add(path.basename(cleanFm));
            }
          }
        }
      }
    } catch (e) {
      // ignore
    }

    if (isMatch) {
      if (userPrompts.length > 0) {
        title = userPrompts[0];
        if (title.length > 60) {
          title = title.substring(0, 57) + '...';
        }
      }

      const agentsMd = path.join(path.resolve(projectRootStr), 'AGENTS.md');
      const ruleFiles = fs.existsSync(agentsMd) ? [agentsMd.replace(/\\/g, '/')] : [];

      const sessionData = {
        vendor: 'antigravity',
        session_id: convId,
        title: title,
        project_root: projectRoot,
        is_active: true,
        created_at: created,
        last_active_at: lastModified,
        log_path: logFile.replace(/\\/g, '/'),
        rule_files: ruleFiles
      };

      if (inspectActivity) {
        sessionData.recent_prompts = userPrompts.length > 5 ? userPrompts.slice(-5) : userPrompts;
        sessionData.recent_touched_files = Array.from(touchedFiles).sort().slice(0, 15);
      }

      sessions.push(sessionData);
    }
  }

  sessions.sort((a, b) => (b.last_active_at || '').localeCompare(a.last_active_at || ''));
  return sessions;
}

function main() {
  const args = process.argv.slice(2);
  const projectRoot = args[0] && !args[0].startsWith('-') ? args[0] : '.';
  const inspect = args.includes('--inspect') || args.includes('-i');

  const sessions = scanAgySessions(projectRoot, null, inspect);
  console.log(JSON.stringify(sessions, null, 2));
}

if (require.main === module) {
  main();
}

module.exports = { scanAgySessions, normalizePath };
